### ISSUE TITLE
emacs (efricas) won't start fricas

### ISSUE THREAD

**Issue by MrBlueXav** [fricas/fricas#83]
When starting efricas, I've got an error message in emacs : 
_fricas-run: wrong number of arguments ( 3 . 3 ) , 4_
My emacs version is 28.1 in Archlinux (Manjaro).
Fricas 1.3.7 is prebuilt in the distribution.

Regards,
Xavier

**Comment by MrBlueXav** [comment]
I found a dirty fix.
The problem is in fricas.el file at line 312 where _fricas-run_ is defined.
_start-process-shell-command_ expects 3 arguments and is given 4 because of "-emacs".
So it works with : 
 **(start-process-shell-command "fricas" (current-buffer) "fricas -emacs"))**
but _fricas-run-command_ is lost. I don't know how to improve this as I'm not a lisp user.
Hope it helps !
Xavier

**Comment by hemmecke** [comment]
Unfortunately, I have no emacs 28.1 and no Manjaro and you probably meant line 325
https://github.com/fricas/fricas/blob/master/contrib/emacs/fricas.el#L325 .

However, I somewhat doubt that you have found the problem. If (in emacs) put your cursor on start-process-shell-command and type C-h f RET, I see at the bottom of this function documentation the following:
```
An old calling convention accepted any number of arguments after COMMAND,
which were just concatenated to COMMAND.  This is still supported but strongly
discouraged.
```
Maybe in Emacs 28.1 it now really only accepts 3 arguments without this "old calling convention". What does the documentation say in your Emacs?

**Comment by hemmecke** [comment]
Oooops... I just found
https://www.masteringemacs.org/article/whats-new-in-emacs-28-1
where it says...
```
'start-process-shell-command' and 'start-file-process-shell-command'
do not support the old calling conventions any longer.
```
So we must change the calling accordingly.

**Comment by hemmecke** [comment]
efricas relies on calling "fricas -emacs". In my eyes, it would make sense to change FRICASCMD in bin/efricas to
```
FRICASCMD='...../fricas -emacs'
```
i.e. add `-emacs` directly there. Furthermore in your installed fricas.el you should remove "-emacs" from line 327.
https://github.com/fricas/fricas/blob/master/contrib/emacs/fricas.el#L327
And finally, in line 130
https://github.com/fricas/fricas/blob/master/contrib/emacs/fricas.el#L130
change it to
```
(or (getenv "FRICASCMD") "fricas -emacs")
```
Please report back if it works for you so that we can put it into the repo.

By the way, do you know https://bitbucket.org/pdo/axiom-environment.git ?
Despite of its name, it works perfectly with FriCAS and also allows you to execute fricas commands inside Emacs.

**Comment by MrBlueXav** [comment]
Yes it works for me, I made all changes.
Thank you !
I'll check axiom-environment.
I'd prefer using jupyter but couldn't make it work...

**Comment by hemmecke** [comment]
Can you contact me via email. I'd like to know more about your problems with jfricas.

**Comment by hebisch** [comment]
On Sat, Apr 23, 2022 at 12:57:28PM -0700, Ralf Hemmecke wrote:
> efricas relies on calling "fricas -emacs". In my eyes, it would make sense to change FRICASCMD in bin/efricas to
> ```
> FRICASCMD='...../fricas -emacs'
> ```
> i.e. add `-emacs` directly there. Furthermore in your installed fricas.el you should remove "-emacs" from line 327.
> https://github.com/fricas/fricas/blob/master/contrib/emacs/fricas.el#L327
> And finally, in line 130
> https://github.com/fricas/fricas/blob/master/contrib/emacs/fricas.el#L130
> change it to
> ```
> (or (getenv "FRICASCMD") "fricas -emacs")
> ```
> Please report back if it works for you so that we can put it into the repo.

This is good test but I do not think it should go to repo.  Rather,
we should explicitely concatenate strings like (untested):

  (start-process-shell-command "fricas" (current-buffer)
                               (concat fricas-run-command
                                       " -emacs"))

Note: When you put this in the function you need extra closing parenthesis.

-- 
                              Waldek Hebisch

**Comment by hemmecke** [comment]
What conceptual reason makes your suggestion better than mine, except for "I do not think"?
Although unlikely, someone could want a separate script "myfricasemacs".

**Comment by hebisch** [comment]
On Sun, Apr 24, 2022 at 12:04:55AM -0700, Ralf Hemmecke wrote:
> What conceptual reason makes your suggestion better than mine, except for "I do not think"?

OK, saying the obvious:

a) What I suggest is probably minimal change.  Why do changes that
   gain us nothing?
b) Related to a: this is backward compatible change
c) Limit number of places that must know about '-emacs'
d) In most likely use case, that is setting non standard paths makes
   setting simpler

> Although unlikely, someone could want a separate script "myfricasemacs".

Well, ignoring extra argument requires essentially no work, so
your suggestion does not really makes things easier even in such
use case.

-- 
                              Waldek Hebisch

**Comment by hemmecke** [comment]
Fixed with commit 1290e5c83b147202ecff5db6d67cbc2377ce09ff.
